src/Symfony/Component/HttpClient/CachingHttpClient.php

```diff
@@ -0,0 +1,114 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpClient\Response\MockResponse;
+use Symfony\Component\HttpClient\Response\ResponseStream;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\HttpCache\HttpCache;
+use Symfony\Component\HttpKernel\HttpCache\StoreInterface;
+use Symfony\Component\HttpKernel\RealHttpKernel;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Symfony\Contracts\HttpClient\ResponseStreamInterface;
+
+/**
+ * An caching HTTP client.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class CachingHttpClient extends MockHttpClient
+{
+    use HttpClientTrait;
+
+    private $client;
+    private $cache;
+    private $defaultOptions = self::OPTIONS_DEFAULTS;
+
+    public function __construct(HttpClientInterface $client, StoreInterface $store, array $defaultOptions = [], LoggerInterface $logger = null)
+    {
+        $this->client = $client;
+        $kernel = new RealHttpKernel($client, $logger);
+        $this->cache = new HttpCache($kernel, $store, $defaultOptions);
+
+        unset($defaultOptions['debug']);
+        unset($defaultOptions['default_ttl']);
+        unset($defaultOptions['private_headers']);
+        unset($defaultOptions['allow_reload']);
+        unset($defaultOptions['allow_revalidate']);
+        unset($defaultOptions['stale_while_revalidate']);
+        unset($defaultOptions['stale_if_error']);
+
+        if ($defaultOptions) {
+            [, $this->defaultOptions] = self::prepareRequest(null, null, $defaultOptions, self::OPTIONS_DEFAULTS);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        [$url, $options] = $this->prepareRequest($method, $url, $options, $this->defaultOptions, true);
+        $url = implode('', $url);
+
+        if (!empty($options['body']) || !\in_array($method, ['GET', 'HEAD', 'OPTIONS', 'TRACE'])) {
+            return $this->client->request($method, $url, $options);
+        }
+
+        $request = Request::create($url, $method);
+        $request->attributes->set('http_client_options', $options);
+        $request->headers->add($options['headers']);
```

Comment:
You also need to add support for Cookies.

Reply:
fixed